client/src/components/Login/loginForm/index.jsx

```diff
@@ -1,83 +1,95 @@
 import React, { Component } from 'react';
 import { Redirect } from 'react-router-dom';
+import { Icon, Spin } from 'antd';
 import axios from 'axios';
+
 import image from './user.png';
-import { Icon } from 'antd';
 import './style.css';
 
 export default class Login extends Component {
   state = {
     user_name: '',
     password: '',
     loginError: '',
-    redirect: ''
+    redirect: '',
+    loading: false
   };
 
   handleChange = ({ target: { name, value } }) => {
     this.setState({ [name]: value.trim() });
   };
+  
   login = event => {
-    event.preventDefault();
     const { password, user_name } = this.state;
-    const { setUserInfo } = this.props;
+    const { setUserInfo, location } = this.props;
+
+    event.preventDefault();
+
     if (!password || !user_name) {
-      this.setState({ loginError: 'Fill All Fields' });
+      this.setState({ loginError: 'You should fill all fields' });
     } else {
-      this.setState({ loginError: '' });
+      this.setState({ loginError: '', loading: true });
       axios
-        .post('/api/v1/login', {
-          user_name: user_name,
-          password: password
-        })
+      .post('/api/v1/login', {
+        user_name,
+        password
+      })
         .then(result => {
           if (result.status === 200) {
-            this.props.location.state.from
-              ? this.setState({
-                  redirect: this.props.location.state.from.pathname
-                })
-              : this.setState({ redirect: '/' });
+            // I am not sure about redirecting here, because line 45 changes the App.js state which cause redirect
```

Comment:
@ashatat It's tough for me to say with 100% confidence without running the app, but I *think* you still need the redirect here, because in App.js the state will change, but it doesn't look like App.js will do any redirecting as a result of that change
Also I think that's fine, to have this component be responsible for the redirecting (and similarly if there's a logout button, the handler for that button can be responsible for doing a redirect).
Also if it makes things simpler for you, you could call this.props.history.push('/') instead of having the <Redirect> component in this render function.

Reply:
I think because I check Authentication each time inside the LogedoutRoute, so when the App render function gets called it'll call the LogedoutRoute and it'll check Authentication which cases redirect which will unmount the LoginComponent before that redirect happens
but I think if I use `this.props.history.push()` then it'll redirect before unmount will happen
:confused: